dimod/include/dimod/constrained.h

```diff
@@ -0,0 +1,320 @@
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#pragma once
+
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "dimod/abc.h"
+#include "dimod/quadratic_model.h"
+#include "dimod/vartypes.h"
+
+namespace dimod {
+
+enum Sense { LE, GE, EQ };
+
+template <class Bias, class Index>
+class ConstrainedQuadraticModel;
+
+template <class Bias, class Index>
+class Constraint : public abc::QuadraticModelBase<Bias, Index> {
+ public:
+    /// The first template parameter (Bias).
+    using bias_type = Bias;
+
+    /// The second template parameter (Index).
+    using index_type = Index;
+
+    /// Unsigned integral type that can represent non-negative values.
+    using size_type = std::size_t;
+
+    using base_type = abc::QuadraticModelBase<bias_type, index_type>;
+
+    using parent_type = ConstrainedQuadraticModel<bias_type, index_type>;
+
+ private:
+    const parent_type* parent_;
+
+    std::vector<index_type> variables_;
+    Sense sense_;
+    bias_type rhs_;
+
+    std::unordered_map<index_type, index_type> indices_;
+
+ public:
+    Constraint(const parent_type* parent, Sense sense)
+            : base_type(), parent_(parent), variables_(), sense_(sense), rhs_(0) {}
+
+    Constraint(const parent_type* parent, std::vector<index_type>&& variables,
+               std::vector<bias_type>&& biases, Sense sense, bias_type rhs)
+            : base_type(std::move(biases)),
+              parent_(parent),
+              variables_(std::move(variables)),
+              sense_(sense),
+              rhs_(rhs) {
+        index_type i = 0;
+        for (const index_type& v : this->variables_) {
+            this->indices_[v] = i++;
+        }
+    }
+
+    Constraint(const parent_type* parent, const Constraint& constraint)
+            : base_type(constraint),
+              parent_(parent),
+              variables_(constraint.variables_),
+              sense_(constraint.sense_),
+              rhs_(constraint.rhs_),
+              indices_(constraint.indices_) {}
+
+    Constraint(Constraint&& other) noexcept { *this = std::move(other); }
+
+    Constraint& operator=(Constraint&& other) noexcept {
+        if (this != &other) {
+            base_type::operator=(std::move(other));
+            this->parent_ = other.parent_;
+            this->variables_ = std::move(other.variables_);
+            this->sense_ = other.sense_;
+            this->rhs_ = other.rhs_;
+            this->indices_ = std::move(other.indices_);
+        }
+        return *this;
+    }
+
+    void add_linear(index_type v, bias_type bias) {
+        base_type::add_linear(this->add_variable(v), bias);
+    }
+
+    void add_quadratic(index_type u, index_type v, bias_type bias) {
+        base_type::add_quadratic(this->add_variable(u), this->add_variable(v), bias);
```

Comment:
In the c++ side of cqm, the function add_quadratic is called with the internal indices, the mapping is sorted out outside. I do not know which one is better. Here the mapping seems to be stored inside. The advantage of the c++ version is that when you are going over all the constraints related to a variable, you have better cache locality when reading the map, but I think this is not important here, so probably this is fine.

See here 

https://git.dwavesys.local/projects/TA/repos/cqm_solver/browse/cqmsolver/cppcqm/quadratic_model.h#1525